src/fft.rs

```diff
@@ -0,0 +1,299 @@
+        /// `output_im` - A mutable slice to store the imaginary parts of the FFT output.
+        ///
+        /// # Panics
+        ///
+        /// Panics if `output_re.len() != output_im.len()` and `input_re.len()` == `output_re.len()`
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use phastft::fft::{r2c_fft_f32, r2c_fft_f64};
+        ///
+        /// let big_n = 16;
+        /// let input: Vec<f64> = (1..=big_n).map(|x| x as f64).collect();
+        /// let mut output_re = vec![0.0; big_n];
+        /// let mut output_im = vec![0.0; big_n];
+        /// r2c_fft_f64(&input, &mut output_re, &mut output_im);
+        ///
+        /// let input: Vec<f32> = (1..=big_n).map(|x| x as f32).collect();
+        /// let mut output_re: Vec<f32> = vec![0.0; 16];
+        /// let mut output_im: Vec<f32> = vec![0.0; 16];
+        /// r2c_fft_f32(&input, &mut output_re, &mut output_im);
+        /// ```
+        /// # References
+        ///
+        /// This implementation is based on the concepts discussed in
+        /// [Levente Kovács' post](https://kovleventer.com/blog/fft_real/).
+        pub fn $func_name(
+            input_re: &[$precision],
+            output_re: &mut [$precision],
+            output_im: &mut [$precision],
+        ) {
+            assert!(output_re.len() == output_im.len() && input_re.len() == output_re.len());
+            let big_n = input_re.len();
+
+            // Splitting odd and even
+            let (mut z_even, mut z_odd): (Vec<_>, Vec<_>) =
+                input_re.chunks_exact(2).map(|c| (c[0], c[1])).unzip();
+
+            // Z = np.fft.fft(z)
+            $fft_func(&mut z_even, &mut z_odd, Direction::Forward);
+
+            let mut z_x_re = Vec::with_capacity(z_even.len());
+            let mut z_x_im = Vec::with_capacity(z_even.len());
+            let mut z_y_re = Vec::with_capacity(z_even.len());
+            let mut z_y_im = Vec::with_capacity(z_even.len());
+
+            z_x_re.push(0.0);
+            z_x_im.push(0.0);
+            z_y_re.push(0.0);
+            z_y_im.push(0.0);
+
+            // Zminconj = np.roll(np.flip(Z), 1).conj()
+            // Zx =  0.5  * (Z + Zminconj)
+            // Zy = -0.5j * (Z - Zminconj)
+            z_even
+                .iter()
+                .skip(1)
+                .zip(z_odd.iter().skip(1))
+                .zip(z_even.iter().skip(1).rev())
+                .zip(z_odd.iter().skip(1).rev())
+                .for_each(|(((z_e, z_o), z_e_mc), z_o_mc)| {
+                    let a = *z_e;
+                    let b = *z_o;
+                    let c = *z_e_mc;
+                    let d = -(*z_o_mc);
+
+                    let t = 0.5 * (a + c);
+                    let u = 0.5 * (b + d);
+                    let v = -0.5 * (a - c);
+                    let w = 0.5 * (b - d);
+
+                    z_x_re.push(t);
+                    z_x_im.push(u);
+                    z_y_re.push(w);
+                    z_y_im.push(v);
```

Comment:
Pushing in a loop isn't great, because it bumps 4 counters on every loop iteration, and probably involves a branch on capacity too.

If we could instead iterate over all of these inputs in a chunked way, with `chunks_exact`, write into intermediate buffers of type `scratch_space = [$precision; 4]` (or any other reasonable size) and then `.extend_from_slice(scratch_space)` on each vector, we might save ourselves half the branches, and unlock some SIMD opportunities too now that we're operating on batches of values.

Or we might invoke LLVM's wrath with non-inlined calls to `memcpy()` for anything over `[f32; 4]` or `[f64; 2]` - dunno. Gotta try it and see.